File: src/utils/twitch.ts
/**
 * Twitch URL/ID extraction utilities
 */

export type TwitchTarget =
  | { type: "channel"; login: string }
  | { type: "vod"; vodId: string }
  | { type: "clip"; clipId: string };

/**
 * Extract Twitch target from various URL formats or raw input.
 * Supports:
 * - twitch.tv/{login}
 * - twitch.tv/videos/{id}
 * - twitch.tv/{login}/video/{id}
 * - twitch.tv/{login}/clip/{slug}
 * - clips.twitch.tv/{slug}
 * - Raw username (1-25 chars, alphanumeric + underscore)
 * - Raw numeric ID prefixed with "v" → VOD
 */
export function extractTwitchTarget(input: string): TwitchTarget | null {
  const trimmed = input.trim();
  if (!trimmed) return null;

  // Try URL patterns first
  try {
    // Normalize: add protocol if missing
    let urlStr = trimmed;
    if (!urlStr.startsWith("http://") && !urlStr.startsWith("https://")) {
      if (urlStr.startsWith("twitch.tv") || urlStr.startsWith("www.twitch.tv") || urlStr.startsWith("clips.twitch.tv")) {
        urlStr = "https://" + urlStr;
      }
    }

    if (urlStr.startsWith("http")) {
      const url = new URL(urlStr);
      const host = url.hostname.replace("www.", "");

      // clips.twitch.tv/{slug}
      if (host === "clips.twitch.tv") {
        const slug = url.pathname.split("/").filter(Boolean)[0];
        if (slug) return { type: "clip", clipId: slug };
      }

      if (host === "twitch.tv") {
        const parts = url.pathname.split("/").filter(Boolean);

        // /videos/{id}
        if (parts[0] === "videos" && parts[1] && /^\d+$/.test(parts[1])) {
          return { type: "vod", vodId: parts[1] };
        }

        // /{login}/video/{id}
        if (parts.length >= 3 && parts[1] === "video" && /^\d+$/.test(parts[2])) {
          return { type: "vod", vodId: parts[2] };
        }

        // /{login}/clip/{slug}
        if (parts.length >= 3 && parts[1] === "clip" && parts[2]) {
          return { type: "clip", clipId: parts[2] };
        }

        // /{login} (but not reserved paths)
        const reserved = new Set(["directory", "downloads", "jobs", "settings", "videos", "search", "p"]);
        if (parts.length >= 1 && parts[0] && !reserved.has(parts[0].toLowerCase())) {
          const login = parts[0].toLowerCase();
          if (/^[a-zA-Z0-9_]{1,25}$/.test(login)) {
            return { type: "channel", login };
          }
        }
      }
    }
  } catch {
    // Not a valid URL, try raw patterns below
  }

  // Raw VOD ID with "v" prefix (e.g., "v2345678901")
  if (/^v\d{1,12}$/.test(trimmed)) {
    return { type: "vod", vodId: trimmed.slice(1) };
  }

  // Raw numeric ID → VOD (only if clearly numeric and long enough to be a VOD)
  if (/^\d{7,12}$/.test(trimmed)) {
    return { type: "vod", vodId: trimmed };
  }

  // Raw username (1-25 alphanumeric + underscore, must start with letter)
  if (/^[a-zA-Z][a-zA-Z0-9_]{0,24}$/.test(trimmed)) {
    return { type: "channel", login: trimmed.toLowerCase() };
  }

  return null;
}

/**
 * Extract Twitch channel login from a Job object.
 * Tries job.url first, then falls back to channelName.
 */
export function extractTwitchLoginFromJob(job: { url?: string; channelName?: string }): string | null {
  if (job.url) {
    const match = job.url.match(/twitch\.tv\/([a-zA-Z0-9_]+)/);
    if (match) return match[1].toLowerCase();
  }
  if (job.channelName && job.channelName !== "Unknown") {
    return job.channelName.toLowerCase();
  }
  return null;
}

File: internal/utils/twitch.go
package utils

import (
	"net/url"
	"regexp"
	"strings"
)

// TwitchTargetType represents the type of Twitch content.
type TwitchTargetType string

const (
	TwitchChannel TwitchTargetType = "channel"
	TwitchVOD     TwitchTargetType = "vod"
	TwitchClip    TwitchTargetType = "clip"
)

// TwitchTarget represents a parsed Twitch URL or identifier.
type TwitchTarget struct {
	Type  TwitchTargetType
	Value string // login name, VOD ID, or clip slug
}

var (
	// twitchLoginRegex matches a valid Twitch login name. Per Twitch's
	// account creation rules: starts with a letter, then up to 24 more
	// letters / digits / underscores (25-char total cap). Case-insensitive
	// is documented but accounts are stored lowercase server-side; callers
	// should ToLower before comparing for equality.
	twitchLoginRegex  = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_]{0,24}$`)
	twitchVODPrefixRe = regexp.MustCompile(`^v(\d{1,12})$`)
	twitchVODBareRe   = regexp.MustCompile(`^\d{7,12}$`)
	// twitchVODIDRe validates a VOD ID taken from a URL path. URL context
	// already disambiguates it from other numbers, so unlike twitchVODBareRe
	// it accepts short legacy IDs (< 7 digits).
	twitchVODIDRe = regexp.MustCompile(`^\d{1,12}$`)
	// twitchClipSlugRe validates a clip slug ("AwkwardSalamander-271WrMRkSrlpFvOY"
	// style or legacy CamelCase). Rejects empty and path-garbage values so a
	// malformed URL surfaces as "unrecognized input" instead of a job that
	// fails downstream with an empty/garbled ID.
	twitchClipSlugRe = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
	// twitchReservedPaths lists first-path-segment slugs on twitch.tv that
	// are site navigation / product pages rather than channel logins, so we
	// must not treat them as channel names when parsing /<slug>.
	twitchReservedPaths = map[string]bool{
		"directory":     true,
		"downloads":     true,
		"jobs":          true,
		"settings":      true,
		"videos":        true,
		"search":        true,
		"p":             true, // /p/... promotional/product pages
		"store":         true,
		"tag":           true, // legacy tag browsing
		"tags":          true,
		"turbo":         true,
		"drops":         true,
		"wallet":        true,
		"prime":         true,
		"subs":          true,
		"subscriptions": true,
		"friends":       true,
		"inventory":     true,
		"payments":      true,
	}
)

// ExtractTwitchTarget extracts a Twitch target from a URL or identifier.
func ExtractTwitchTarget(input string) *TwitchTarget {
	input = strings.TrimSpace(input)
	if input == "" {
		return nil
	}

	// Normalize bare twitch.tv/... input without protocol
	if strings.HasPrefix(input, "twitch.tv/") || strings.HasPrefix(input, "www.twitch.tv/") || strings.HasPrefix(input, "clips.twitch.tv/") {
		input = "https://" + input
	}

	// Try parsing as URL
	u, err := url.Parse(input)
	if err == nil {
		host := strings.ToLower(u.Hostname())
		if host == "www.twitch.tv" || host == "twitch.tv" || host == "m.twitch.tv" {
			return parseTwitchURL(u)
		}
		// clips.twitch.tv/{slug} — first path segment only
		if host == "clips.twitch.tv" {
			slug := strings.TrimPrefix(u.Path, "/")
			slug, _, _ = strings.Cut(slug, "/")
			if twitchClipSlugRe.MatchString(slug) {
				return &TwitchTarget{Type: TwitchClip, Value: slug}
			}
		}
	}

	// Try as v-prefixed VOD ID (v123456)
	if matches := twitchVODPrefixRe.FindStringSubmatch(input); matches != nil {
		return &TwitchTarget{Type: TwitchVOD, Value: matches[1]}
	}

	// Try as bare numeric VOD ID (7-12 digits to distinguish from other numbers)
	if twitchVODBareRe.MatchString(input) {
		return &TwitchTarget{Type: TwitchVOD, Value: input}
	}

	// Try as bare login (must start with letter)
	if twitchLoginRegex.MatchString(input) {
		return &TwitchTarget{Type: TwitchChannel, Value: strings.ToLower(input)}
	}

	return nil
}

func parseTwitchURL(u *url.URL) *TwitchTarget {
	path := strings.TrimPrefix(u.Path, "/")
	parts := strings.Split(path, "/")

	if len(parts) == 0 || parts[0] == "" {
		return nil
	}

	// /videos/123456
	if parts[0] == "videos" && len(parts) >= 2 {
		if twitchVODIDRe.MatchString(parts[1]) {
			return &TwitchTarget{Type: TwitchVOD, Value: parts[1]}
		}
		return nil
	}

	// /channel/clip/slug
	if len(parts) >= 3 && parts[1] == "clip" {
		if twitchClipSlugRe.MatchString(parts[2]) {
			return &TwitchTarget{Type: TwitchClip, Value: parts[2]}
		}
		return nil
	}

	// /channel/video/123456
	if len(parts) >= 3 && parts[1] == "video" {
		if twitchVODIDRe.MatchString(parts[2]) {
			return &TwitchTarget{Type: TwitchVOD, Value: parts[2]}
		}
		return nil
	}

	// /channel (skip reserved paths)
	login := strings.ToLower(parts[0])
	if twitchReservedPaths[login] {
		return nil
	}
	if twitchLoginRegex.MatchString(login) {
		return &TwitchTarget{Type: TwitchChannel, Value: login}
	}

	return nil
}
